Fix bug 1486500 - Make the Translate UI localizable.

adngdb opened the pull request:
Pontoon in French, ladies and gentlemen! :sparkles: 

I think this one requires extensive review. I have never done this before, and it's an area that is nothing near new, so there are others who have a lot more experience than I do. I thus fear I am making a lot of mistakes with my approach.

There are also a lot of decisions made here. Not that they cannot be changed later, but we should at least identify which ones we know will have to evolve. Among those decisions: 

- where to put localization files (`/frontend/public/static/locale/${locale}/`). I believe that should change and be pushed to a different repo, so that we do not pollute our commit history with translations ([mozilla/addons-frontend](https://github.com/mozilla/addons-frontend) does it and it puzzles me).
- how to do language negotiation. I made the decision to do it in the server, so that we can benefit from user settings. Right now, the `preferred homepage` setting takes precedence over everything else (which makes this easy to test, just changed your team page to `fr`). It doesn't necessarily makes sense, as users might work primarily on a locale but prefer to see the UI in another one (for example, I expect some people, myself included, will want to keep the UI in `en-US`).
- what locales to fall back to. Right now, there is only one fallback, and that's the default content from the code (not even the content of the `en-US` file). That is very likely wrong, but I don't know what it should actually be.
- what IDs to use for strings. I'm not sure we need to have strict rules, but my instinct tells me we should. What do Firefox developers do? Other Fluent users? And is the rule I proposed good? 

I'm also concerned that I might be using Fluent incorrectly in some cases.

For all these reasons, I'd like to request review from @mathjazz, @stas and @Pike. Anyone else interested in this conversation is very welcome to join.

adngdb commented:
The current build error is unrelated to the logic in the code, please disregard it. Is a compatibility issue with `fluent/compat` and our build system. I'll fix it as soon as I can next week.

mathjazz commented:
I'm unable to `make build`:

```make
Step 24/27 : RUN cd /app/frontend/ && yarn build
 ---> Running in 16f5754687cd
yarn run v1.10.1
$ react-scripts build
Creating an optimized production build...
Failed to compile.

Failed to minify the code from this file: 

 	./node_modules/fluent-react/compat.js:26 

Read more here: http://bit.ly/2tRViJ9

error Command failed with exit code 1.
info Visit https://yarnpkg.com/en/docs/cli/run for documentation about this command.
ERROR: Service 'webapp' failed to build: The command '/bin/sh -c cd /app/frontend/ && yarn build' returned a non-zero code: 1
make: *** [build] Error 1
Leopold:pontoon mathjazz$ 
```

stasm commented:
Ah yes, you'll need to update to `create-react-app` 2.x which uses a newer version of `uglify-es`. I think they also transpile `node_modules`, too, which should make it possible to also drop the `fluent-react/compat` build in favor of the vanilla `fluent-react`.

adngdb commented on frontend/public/static/locale/fr/translate.ftl:
```diff
@@ -0,0 +1,39 @@
+plural-selector-form-two = deux
+plural-selector-form-few = un peu
+plural-selector-form-many = beaucoup
+plural-selector-form-other = autre
```
I understand why you would want to keep these in English, but I also think that can be applied to most of the rest of the application (as we use domain-specific language in many places). I believe the whole point of this is to be able to translate strings like these to make them clearer to translators. 

What's your rationale for not translating them?

stasm replied:
(The following is my own interpretation. We might want localizers' feedback on this because ultimately, this UI is supposed to help them create better translations.)

These names are symbolic categories, not actual numerals. They just happen to be called after English numerals. For instance, translating `one` as `un` might be misleading because this category is also used with some decimals, e.g. _1,5 jour_. Other languages may have more complex rules. For instance in [Russian](http://www.unicode.org/cldr/charts/29/supplemental/language_plural_rules.html#ru), `one` is used for numbers which end with `1` but not with `11`. Translating the name of this category to `один` may suggest that it's only to be used with the number `1`.

Pike replied:
The culprit here is the usage of technical jargon in the UI. `two` isn't any more indicative than `Zwei` that this is meant to be a dual form. Or that `one` means _Singular_.

But .... replacing the short keys with actual grammatical Terms for the target language might lead to UI regressions, which we shouldn't introduce as part of this transition, I think.

My recommendation would be to not localize these terms now, and to file a follow-up to create a UI exposing the correct grammatical forms later on. And those human-targetting strings should then also be localizable.

mathjazz replied:
Note that we use two different UIs for plurals.

The one we talk about here is used for Gettext translations and works better with Machinery (and the rest of the UI). The other one is used by Fluent and allows (will allow) for features only supported in Fluent (e.g. custom variants, default variants, having sibling elements).

I agree it's best to not localize this for now.

adngdb replied:
Thanks everyone for your inputs! I shall not make these localizable for now. :-)

adngdb commented on frontend/README.md:
```diff
@@ -142,6 +142,45 @@ it('does something', () => {
+
+The second thing you need to do is to add that new string into our main translation files. Translations available to localizers are based on those files, so if you forget to put the new content there, no one will be able to translate that content.
+
+The source language for our application is `en-US` (that is what you should use by default when writing content in the code, and in the language files). You can find them in `frontend/public/static/locale/en-US/`. We currently have only one file, named `translate.ftl`, and that is where all content should be added.
```
I think it would be amazing to be able to have FTL files right next to the file containing the strings, like we do right now with `.css` and `.test.js` files. However, I don't see how that would work with translations collection (for Pontoon for example) and translations storage (because I don't want to have one file per locale in my code).

However, I can imagine some automation that would collect all `.ftl` file from the `src/` folder and merge them into one / some main files, in a locale folder, and that would serve as the translations reference.

stasm replied:
That's a good point. In the future, we might want to build tools to help with the collection step.

For now, keeping everything in a single file will make things much easier. We could consider splitting translations into smaller files before the single file gets _too_ long. I think it would require some kind of an `l10n.toml` project config, however? CC @Pike.

stasm replied:
Ah, nvm, I see that @Pike already wrote a lengthy comment about this.

adngdb replied:
Filed [bug 1504710](https://bugzilla.mozilla.org/show_bug.cgi?id=1504710).

Pike commented:
As for localization files, we sadly don't have best practices, but we have a few popular options.

L10n in tree
---
These are nice for developing with multi-lingual apps in mind. You can just use your in-development tree, and switch the language to try something out.
There are different ways to do this.

1. Files are in the same hierarchy as in en-US.
2. Files are in a separate part of the source repository. project-config supports that, and Fluent doesn't care, if you're able to load files from multiple roots.

There are also different commit policies.

1. pontoon commits directly to master. comes with a ton of noise in commit history.
2. pontoon commits to a branch with an ongoing PR to merge (which can squash)
3. a nightly cron job can commit from an external repo

L10n in separate repo
---
This is nice 'cause the developers are not bothered with the l10n commits and files at all. (Unless you're doing the hybrid which is the nightly cron above.)

This makes a multilingual deployment different to your development setup, and you'll need to make more of an effort to test the post-deployment configuration.

adngdb commented:
@stasm I believe I addressed all your comments.

@Pike I personally hate having l10n commits in my repo. I would be in favor of having all translations in a separate repo, and include them at deploy time (for prod) and build time (for dev). I believe that would not be too hard to do with Docker, and thus we can have a dev environment that also has translations.

Pike commented on frontend/README.md:
```diff
@@ -142,6 +142,45 @@ it('does something', () => {
+
+### Social contract
+
+Fluent uses the concept of a *social contract* between developer and localizers. This contract is established by the selection of a unique identifier, called `l10n-id`, which carries a promise of being used in a particular place to carry a particular meaning.
```
Some of use this phrase, but I'm not happy with _Social Contract_. This is a development pattern, and an API surface to external code and processes that is needed to get continuous localization and change management in general.

I prefer to word this around semantic versioning for strings, with no "middle" version number. If you're breaking backwards compat, use a new ID. If you're not, seriously, don't use a new ID.

I also like to describe the l10n ids as good variable names. Often, if you change the semantics around a given UI, a different name for that variable is part of the change.

stasm commented on frontend/src/core/api/l10n.js:
```diff
@@ -0,0 +1,12 @@
+    async get(locale: string) {
+        const url = this.getFullURL(`/static/locale/${locale}/translate.ftl`);
+        const response = await fetch(url);
+        return await response.text();
```
Playing the devil's advocate: Right now this whole API class adds complexity, too :) You could use the regular `APIBase.fetch` or even the native `fetch`, and build the URL in `core/l10n/actions`.

The decision to abstract this logic into its own class, `L10nAPI`, means to me as a reader of this code, that l10n is special enough to deserve its own API class. This is why I suggested move some l10n-specific logic here, and making the action creator more "stupid".

I think this boils down to what you want `L10nAPI.get` to return. A string, at which point its not very different from a regular `fetch`? Or maybe a `FluentResource`? This is up to you and up to the conventions used by other modules. Technically both are fine and r=me. (You'll note that `addMessages` is just `FluentResource.fromString()` followed by an `addResource`.)

adngdb replied:
I'm going to keep things as they are for now, and next time I refactor the API classes I'll have that in mind.

adngdb commented on frontend/src/core/l10n/actions.js:
```diff
@@ -0,0 +1,77 @@
+            locales.push('en-US');
+        }
+
+        Promise.all(locales.map(locale => {
```
Sweet! I'm not yet very confident with async/await and Promises, so thanks for pointing those out. :-)

adngdb commented on frontend/src/core/l10n/components/AppLocalizationProvider.js:
```diff
@@ -0,0 +1,58 @@
+
+        if (!l10n.bundles.length) {
+            // TODO: Show a loader.
+            return <div>LOADING</div>;
```
There's going to be a loader here, @jotes is working on it, see [bug 1489054](https://bugzilla.mozilla.org/show_bug.cgi?id=1489054). :)

stasm replied:
Ah, that's great. Thanks for pointing me to the bug.

adngdb commented on frontend/src/modules/entitydetails/components/Metadata.js:
```diff
@@ -102,7 +110,9 @@ export default class Metadata extends React.Component<Props> {
-        return <Property title='Context'>{ entity.key }</Property>;
+        return <Property title='Context' l10nId='entitydetails-metadata-context'>
+            { entity.key }
+        </Property>;
```
Well I didn't know that was possible! Thank you!

stasm replied:
The `l10nId` is a common pattern for solving more complex scenarios, so your intuition was good :)

adngdb commented:
Remaining things to address here: 

 - [ ] Update README according to Axel's comment
 - [ ] Consider options for langneg (new user setting or just classic langneg)
 - [ ] Start conversations or bugs about sudo locale, where to put FTL files, and adding Pontoon on Pontoon
